contracts/src/TeleporterTokenSource.sol

```diff
@@ -0,0 +1,138 @@
+// (c) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import {TeleporterMessageInput, TeleporterFeeInfo} from "@teleporter/ITeleporterMessenger.sol";
+import {TeleporterOwnerUpgradeable} from "@teleporter/upgrades/TeleporterOwnerUpgradeable.sol";
+import {ITeleporterTokenBridge} from "./interfaces/ITeleporterTokenBridge.sol";
+import {IWarpMessenger} from
+    "@avalabs/subnet-evm-contracts@1.2.0/contracts/interfaces/IWarpMessenger.sol";
+import {IERC20} from "@openzeppelin/contracts@4.8.1/token/ERC20/IERC20.sol";
+
+/**
+ * THIS IS AN EXAMPLE CONTRACT THAT USES UN-AUDITED CODE.
+ * DO NOT USE THIS CODE IN PRODUCTION.
+ */
+
+/**
+ * @dev Abstract contract for a Teleporter token bridge that sends tokens to {TeleporterTokenDestination} instances.
+ *
+ * This contract also handles multihop transfers, where tokens sent from a {TeleporterTokenDestination}
+ * instance are forwarded to another {TeleporterTokenDestination} instance.
+ */
+abstract contract TeleporterTokenSource is ITeleporterTokenBridge, TeleporterOwnerUpgradeable {
+    // Warp precompile used for sending and receiving Warp messages.
+    address public constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+
+    // The blockchain ID of the chain this contract is deployed on.
+    bytes32 public immutable blockchainID;
+
+    // The ERC20 token this contract uses to pay for Teleporter fees.
+    IERC20 public immutable feeToken;
+
+    // Tracks the balances of tokens sent to other bridge instances.
+    // Bridges are not allowed to unwrap more than has been sent to them.
+    // (destinationBlockchainID, destinationBridgeAddress) -> balance
```

Comment:
Nit: Could you have the doc comments on the state variables be natspec formatted also?